src/Plugins/OrientationAnalysis/src/OrientationAnalysis/Filters/Algorithms/ComputeShapesTriangleGeom.cpp

```diff
@@ -411,49 +386,136 @@ Result<> ComputeShapesTriangleGeom::operator()()
-        axisLengths[(3 * featureId) + 2] = static_cast<float32>(lengths.zLength);
+        axisLengths[0] = static_cast<float32>(lengths.xLength);
+        axisLengths[1] = static_cast<float32>(lengths.yLength);
+        axisLengths[2] = static_cast<float32>(lengths.zLength);
         auto bOverA = static_cast<float32>(lengths.yLength / lengths.xLength);
         auto cOverA = static_cast<float32>(lengths.zLength / lengths.xLength);
-        aspectRatios[2 * featureId] = bOverA;
-        aspectRatios[(2 * featureId) + 1] = cOverA;
+        aspectRatios[0] = bOverA;
+        aspectRatios[1] = cOverA;
       }
+      m_FilterPtr->updateResults(featureId, omega3, axisEulerAngles, axisLengths, aspectRatios);
+    } // end
+  }
+
+  void operator()(const Range& range) const
+  {
+    convert(range.min(), range.max());
+  }
+};
+
+} // namespace
+
+// -----------------------------------------------------------------------------
+ComputeShapesTriangleGeom::ComputeShapesTriangleGeom(DataStructure& dataStructure, const IFilter::MessageHandler& mesgHandler, const std::atomic_bool& shouldCancel,
+                                                     ComputeShapesTriangleGeomInputValues* inputValues)
+: m_DataStructure(dataStructure)
+, m_InputValues(inputValues)
+, m_ShouldCancel(shouldCancel)
+, m_MessageHandler(mesgHandler)
+{
+}
+
+// -----------------------------------------------------------------------------
+ComputeShapesTriangleGeom::~ComputeShapesTriangleGeom() noexcept = default;
+
+void ComputeShapesTriangleGeom::updateResults(int32 featureId, float32 omega3, const std::array<float32, 3>& axisEulerAngles, const std::array<float32, 3>& axisLengths,
+                                              const std::array<float32, 2>& aspectRatios)
+{
+  std::lock_guard<std::mutex> guard(m_ProgressMessage_Mutex);
+  m_FeatureUpdateCount++;
+  auto& omega3sRef = *m_Omega3s;                 // m_DataStructure.getDataRefAs<Float32Array>(m_InputValues->Omega3sArrayPath);
+  auto& axisEulerAnglesRef = *m_AxisEulerAngles; // m_DataStructure.getDataRefAs<Float32Array>(m_InputValues->AxisEulerAnglesArrayPath);
+  auto& axisLengthsRef = *m_AxisLengths;         // m_DataStructure.getDataRefAs<Float32Array>(m_InputValues->AxisLengthsArrayPath);
+  auto& aspectRatiosRef = *m_AspectRatios;       // m_DataStructure.getDataRefAs<Float32Array>(m_InputValues->AspectRatiosArrayPath);
+
+  omega3sRef.setValue(featureId, omega3);
+
+  axisEulerAnglesRef[3 * featureId] = static_cast<float32>(axisEulerAngles[0]);
+  axisEulerAnglesRef[(3 * featureId) + 1] = static_cast<float32>(axisEulerAngles[1]);
+  axisEulerAnglesRef[(3 * featureId) + 2] = static_cast<float32>(axisEulerAngles[2]);
+
+  axisLengthsRef[3 * featureId] = axisLengths[0];
+  axisLengthsRef[(3 * featureId) + 1] = axisLengths[1];
+  axisLengthsRef[(3 * featureId) + 2] = axisLengths[2];
+
+  aspectRatiosRef[2 * featureId] = aspectRatios[0];
+  aspectRatiosRef[(2 * featureId) + 1] = aspectRatios[1];
```

Comment:
This shouldn't have to be in a lock since you are explicitly avoiding data races by updating on feature index and they are never read from in the loop either. I would move these ahead of the lock to help cut down on lock time or even move it back into the end of the parallel feature loop and just call this function for updates.

Reply:
I don't want to write to one of our DataStores from different threads because I am fairly sure that would cause issues especially considering out-of-core.

Reply:
> I don't want to write to one of our DataStores from different threads because I am fairly sure that would cause issues especially considering out-of-core.

The out-of-core should be okay as long as the get/set methods are called (instead of bracket operator) as matthew suggested. If the concern is chunk/cache flushing/miss rates in multithreading, I think we should test it and potentially implement a custom access pattern like this for if out of core is enabled. The locks will cripple performance and needlessly so for regular computation.

Reply:
@nyoungbq I reworked the algorithm a bit. Now I collect all results for each thread, then pass that group of results down to the filter to be written into the final arrays. This should significantly lessen the number of times that the mutex_lock has to be acquired for each thread.